vnext/ReactUWP/Views/Image/ReactImageBrush.cpp

```diff
@@ -0,0 +1,204 @@
+        winrt::CompositionSurfaceBrush surfaceBrush{ GetOrCreateSurfaceBrush() };
+        surfaceBrush.Stretch(ResizeModeToStretch());
+
+        auto compositionBrush{ surfaceBrush.as<winrt::CompositionBrush>() };
+        if (ResizeMode() == ResizeMode::Repeat)
+        {
+          compositionBrush = GetOrCreateEffectBrush(surfaceBrush);
+        }
+
+        // Only switch CompositionBrush if we are switching to/from ResizeMode::Repeat
+        if (ShouldSwitchCompositionBrush())
+        {
+          CompositionBrush(compositionBrush);
+        }
+      }
+    }
+
+    winrt::CompositionEffectBrush ReactImageBrush::GetOrCreateEffectBrush(winrt::CompositionSurfaceBrush const& surfaceBrush)
+    {
+      if (!m_effectBrush)
+      {
+        auto borderEffect{ winrt::make<winrt::implementation::BorderEffect>() };
+        borderEffect.ExtendX(winrt::CanvasEdgeBehavior::Wrap);
+        borderEffect.ExtendY(winrt::CanvasEdgeBehavior::Wrap);
+
+        winrt::CompositionEffectSourceParameter borderEffectSourceParameter{ L"source" };
+        borderEffect.Source(borderEffectSourceParameter);
+
+        auto effectFactory{ winrt::Window::Current().Compositor().CreateEffectFactory(borderEffect) };
+        m_effectBrush = effectFactory.CreateBrush();
+
+        surfaceBrush.HorizontalAlignmentRatio(0.0f);
+        surfaceBrush.VerticalAlignmentRatio(0.0f);
+
+        m_effectBrush.SetSourceParameter(L"source", surfaceBrush);
+      }
+
+      return m_effectBrush;
+    }
+
+    winrt::CompositionSurfaceBrush ReactImageBrush::GetOrCreateSurfaceBrush()
+    {
+      // If it doesn't exist, create it
+      if (!CompositionBrush())
+      {
+        auto loadedSurface{ winrt::LoadedImageSurface::StartLoadFromUri(SourceUri()) };
+
+        loadedSurface.LoadCompleted([this](IInspectable const&, winrt::LoadedImageSourceLoadCompletedEventArgs const& args)
+        {
+          if (args.Status() == winrt::LoadedImageSourceLoadStatus::Success)
+          {
+            m_imageLoaded = true;
+            UpdateCompositionBrush();
+          }
+        });
+
+        auto surfaceBrush{ winrt::Window::Current().Compositor().CreateSurfaceBrush() };
+        surfaceBrush.Surface(loadedSurface);
+
+        return surfaceBrush;
+      }
+
+      auto surfaceBrush{ CompositionBrush().try_as<winrt::CompositionSurfaceBrush>() };
+
+      // If ResizeMode is set to Repeat, then we're using a CompositionEffectBrush.
+      // Get the CompositionSurfaceBrush from its source.
+      if (!surfaceBrush)
+      {
+        auto effectBrush{ CompositionBrush().as<winrt::CompositionEffectBrush>() };
+        surfaceBrush = effectBrush.GetSourceParameter(L"source").as<winrt::CompositionSurfaceBrush>();
+      }
+
+      return surfaceBrush;
+    }
+
+    bool ReactImageBrush::ShouldSwitchCompositionBrush()
+    {
+      bool effectToSurfaceBrushSwitch{
+        ResizeMode() != ResizeMode::Repeat &&
+        !CompositionBrush().try_as<winrt::CompositionSurfaceBrush>() };
```

Comment:
>!CompositionBrush().try_as<winrt::CompositionSurfaceBrush>() [](start = 8, length = 60)

This check feels a bit awkward - we're looking to see if we have an effect brush set, yes?  In this case can you check for a non-null m_effectBrush instead?

Reply:
I'm caching the effect brush in case they have some kind of mechanism to switch an image's resizeMode at runtime, so we can't just check to see if it's not null. What I want to know is if the current composition brush is not a CompositionSurfaceBrush (because that's when we would want to switch brushes). I've added a "UsingSurfaceBrush()" helper function.